sdk/identity/Azure.Identity/src/DeviceCodeCredential.cs

```diff
@@ -0,0 +1,179 @@
+using System.Linq;
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Azure.Identity
+{
+    /// <summary>
+    /// Details of the device code to present to a user to allow them to authenticate through the device code authentication flow.
+    /// </summary>
+    public struct DeviceCode
+    {
+        /// <summary>
+        /// The UTC time when the device code request will expire.
+        /// </summary>
+        public DateTimeOffset ExpiresOn { get; set; }
+
+        /// <summary>
+        /// The message to display to the user to allow them to authenticate through the device code authentication flow.
+        /// </summary>
+        public string Message { get; set; }
+    }
+
+    /// <summary>
+    /// A <see cref="TokenCredential"/> implementation which authenticates a user using the device code flow, and provides access tokens for that user account.  
+    /// For more information on the device code authentication flow see https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/wiki/Device-Code-Flow.
+    /// </summary>
+    public class DeviceCodeCredential : TokenCredential
+    {
+        private IPublicClientApplication _pubApp = null;
+        private HttpPipeline _pipeline = null;
+        private IAccount _account = null;
+        private IdentityClientOptions _options;
+        private string _clientId;
+        private Func<DeviceCode, Task> _deviceCodeCallback;
+
+        /// <summary>
+        /// Protected constructor for mocking
+        /// </summary>
+        protected DeviceCodeCredential()
+        {
+
+        }
+
+        /// <summary>
+        /// Creates a new DeviceCodeCredential which will authenticate users with the specified application.
+        /// </summary>
+        /// <param name="clientId">The client id of the application to which the users will authenticate.</param>
+        /// TODO: need to link to info on how the application has to be created to authenticate users, for multiple applications
+        /// <param name="deviceCodeCallback">The callback to be executed to display the device code to the user</param>
+        public DeviceCodeCredential(string clientId, Func<DeviceCode, Task> deviceCodeCallback)
+            : this(clientId, deviceCodeCallback, null)
+        {
+            
+        }
+
+        /// <summary>
+        /// Creates a new DeviceCodeCredential with the specifeid options, which will authenticate users with the specified application.
+        /// </summary>
+        /// <param name="clientId">The client id of the application to which the users will authenticate</param>
+        /// <param name="options">The client options for the newly created DeviceCodeCredential</param>
+        /// <param name="deviceCodeCallback">The callback to be executed to display the device code to the user</param>
+        public DeviceCodeCredential(string clientId, Func<DeviceCode, Task> deviceCodeCallback, IdentityClientOptions options)
```

Comment:
Should we pass cancellationToken into the callback?

Reply:
It's async because I basically am mirroring of the shape of the callback that MSAL provides.  MSAL doesn't feed the cancellation token to the callback, but I suppose I could remedy this.  I think it would require using an AsyncLocal so I could map the specific GetToken / GetTokenAsync invocation to the callback invocation.

Reply:
You don't need an async local, you can capture cancellationToken into a lamda here https://github.com/Azure/azure-sdk-for-net/pull/7033/files#diff-8a633908cb9d1453aa79fb6c2617d31bR165